apps/emqx_rule_engine/src/emqx_rule_runtime.erl

```diff
@@ -77,48 +85,80 @@ apply_rule(Rule = #{id := RuleID}, Columns, Envs) ->
         %% ignore the errors if select or match failed
         _:Reason = {select_and_transform_error, Error} ->
             ok = emqx_metrics_worker:inc(rule_metrics, RuleID, 'failed.exception'),
-            ?SLOG(warning, #{
-                msg => "SELECT_clause_exception",
-                rule_id => RuleID,
-                reason => Error
-            }),
+            trace_rule_sql(
+                "SELECT_clause_exception",
+                #{
+                    reason => Error
+                },
+                warning
+            ),
             {error, Reason};
         _:Reason = {match_conditions_error, Error} ->
             ok = emqx_metrics_worker:inc(rule_metrics, RuleID, 'failed.exception'),
-            ?SLOG(warning, #{
-                msg => "WHERE_clause_exception",
-                rule_id => RuleID,
-                reason => Error
-            }),
+            trace_rule_sql(
+                "WHERE_clause_exception",
+                #{
+                    reason => Error
+                },
+                warning
+            ),
             {error, Reason};
         _:Reason = {select_and_collect_error, Error} ->
             ok = emqx_metrics_worker:inc(rule_metrics, RuleID, 'failed.exception'),
-            ?SLOG(warning, #{
-                msg => "FOREACH_clause_exception",
-                rule_id => RuleID,
-                reason => Error
-            }),
+            trace_rule_sql(
+                "FOREACH_clause_exception",
+                #{
+                    reason => Error
+                },
+                warning
+            ),
             {error, Reason};
         _:Reason = {match_incase_error, Error} ->
             ok = emqx_metrics_worker:inc(rule_metrics, RuleID, 'failed.exception'),
-            ?SLOG(warning, #{
-                msg => "INCASE_clause_exception",
-                rule_id => RuleID,
-                reason => Error
-            }),
+            trace_rule_sql(
+                "INCASE_clause_exception",
+                #{
+                    reason => Error
+                },
+                warning
+            ),
             {error, Reason};
         Class:Error:StkTrace ->
             ok = emqx_metrics_worker:inc(rule_metrics, RuleID, 'failed.exception'),
-            ?SLOG(error, #{
-                msg => "apply_rule_failed",
-                rule_id => RuleID,
-                exception => Class,
-                reason => Error,
-                stacktrace => StkTrace
-            }),
+            trace_rule_sql(
+                "apply_rule_failed",
+                #{
+                    exception => Class,
+                    reason => Error,
+                    stacktrace => StkTrace
+                },
+                warning
+            ),
             {error, {Error, StkTrace}}
+    after
+        reset_process_trace_metadata(Columns)
     end.
 
+set_process_trace_metadata(RuleID, #{clientid := ClientID}) ->
```

Comment:
we can try to avoid using process dictionary for this?
we can maybe try to check if current rule is being traced or not directly by calling emqx_trace module
if not being traced, log nothing, if it's being traced emqx_trace directly to log it.

Reply:
> we can try to avoid using process dictionary for this?

We can but I don't think that makes the code better. This would mean that we need to pass the `ClientID` to calls that do tracing. I checked other `?TRACE(...)` calls and it seems common to pass tracing meta data by setting it implicitly using `logger:update_process_metadata(...)`. Maybe it is good to not clutter the code more than necessary by adding parameters that are just needed for tracing?

> we can maybe try to check if current rule is being traced or not directly by calling emqx_trace module
if not being traced, log nothing, if it's being traced emqx_trace directly to log it.

If we don't need to do this optimization (e.g., for performance reason), I think it is better to leave what will actually be logged to the logger/tracing framework (logger). All added trace log messages have level debug and if my understanding is correct they will thrown away unless there is a matching trace, or the user has set the global log level to debug. Also remember that since we have now added clientid to the metadata we might want a trace message to show up in a trace even though the rule is not traced.